core/runtime/exception_state.rs

```diff
@@ -0,0 +1,127 @@
+use crate::error::exception_to_err_result;
+use anyhow::Error;
+use std::cell::Cell;
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::rc::Rc;
+
+#[derive(Default)]
+pub(crate) struct ExceptionState {
+  // TODO(nayeemrmn): This is polled in `exception_to_err_result()` which is
+  // flimsy. Try to poll it similarly to `pending_promise_rejections`.
+  dispatched_exception: Cell<Option<v8::Global<v8::Value>>>,
+  dispatched_exception_is_promise: Cell<bool>,
+  pub(crate) pending_promise_rejections:
+    RefCell<VecDeque<(v8::Global<v8::Promise>, v8::Global<v8::Value>)>>,
+  pub(crate) js_build_custom_error_cb:
+    RefCell<Option<Rc<v8::Global<v8::Function>>>>,
+  pub(crate) js_promise_reject_cb:
+    RefCell<Option<Rc<v8::Global<v8::Function>>>>,
+  pub(crate) js_format_exception_cb:
+    RefCell<Option<Rc<v8::Global<v8::Function>>>>,
+}
+
+impl ExceptionState {
+  pub fn destroy(&self) {
```

Comment:
Maybe `reset()` since the struct is still valid after calling this fn

Reply:
This is a pre-teardown method so I cannot guarantee that it can be re-used after calling destroy. I think if we do add a runtime reset function we can rename it and test it as such but I wouldn't want to give that impression as-is.

Reply:
Okay I see - in this case can it accept `self`? The types currently don't really communicate well that it shouldn't be used anymore.

Reply:
Because of the `Rc<...>` we're kind of stuck with `&self`. It's challenging because we need to ensure that all these callbacks are torn down before the snapshot is taken, and we still have a lot of intertwined classes.

I can improve the docs and maybe write a TODO on it. As I continue to tear out JsRealm, we may actually be able to get to a point where we can manually release all the isolate resources before the snapshot is taken. Right now it's pretty ugly.

Reply:
Renamed to "prepare_to_destroy" to make it a little clearer.